File: golang/day02/main.go
package main

import (
	"bufio"
	"os"
	"strconv"
	"strings"
)

var MAX_CUBES = map[string]int{
	"red":   12,
	"green": 13,
	"blue":  14,
}

func load_game(filename string) (map[int]map[string]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	scanner := bufio.NewScanner(file)
	defer func(file *os.File) {
		err := file.Close()
		if err != nil {
			panic(err)
		}
	}(file)
	gameData := make(map[int]map[string]int)
	for scanner.Scan() {
		line := scanner.Text()
		lineSplit := strings.Split(line, ":")
		gameSplit := strings.Split(lineSplit[0], " ")
		gameID, err := strconv.Atoi(gameSplit[1])
		if err != nil {
			return nil, err
		}
		maxCubes := map[string]int{
			"red":   0,
			"green": 0,
			"blue":  0,
		}
		for _, draw := range strings.Split(lineSplit[1], ";") {
			for _, cube := range strings.Split(draw, ",") {
				cubeSplit := strings.Split(strings.TrimSpace(cube), " ")
				cubeColor := cubeSplit[1]
				cubeCount, err := strconv.Atoi(cubeSplit[0])
				if err != nil {
					return nil, err
				}
				maxCubes[cubeColor] = max(maxCubes[cubeColor], cubeCount)
			}
		}
		gameData[gameID] = maxCubes

	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return gameData, nil
}

func solutionPart1() {
	gameData, err := load_game("input.txt")
	if err != nil {
		panic(err)
	}
	possibleGamesSum := 0
	for gameIndex, game := range gameData {
		if game["red"] <= MAX_CUBES["red"] &&
			game["green"] <= MAX_CUBES["green"] &&
			game["blue"] <= MAX_CUBES["blue"] {
			possibleGamesSum += gameIndex
		}
	}
	println(possibleGamesSum)
}

func solutionPart2() {
	gameData, err := load_game("input.txt")
	if err != nil {
		panic(err)
	}
	gamePowerSum := 0
	for _, game := range gameData {
		gamePowerSum += game["red"] * game["blue"] * game["green"]
	}
	println(gamePowerSum)
}

func main() {
	solutionPart1()
	solutionPart2()
}

File: Python/day02/main.py
from enum import StrEnum, auto


class CubeColor(StrEnum):
    RED = auto()
    GREEN = auto()
    BLUE = auto()


MAX_CUBES = {
    CubeColor.RED: 12,
    CubeColor.GREEN: 13,
    CubeColor.BLUE: 14
}


def load_game(filename: str) -> dict[int, dict[CubeColor, int]]:
    game_data = {}
    with open(filename) as f:
        for line in f:
            game, draws = line.split(":")
            _, game_id = game.split(" ")
            game_data[int(game_id)] = {}
            cubes_max = {
                CubeColor.RED: 0,
                CubeColor.GREEN: 0,
                CubeColor.BLUE: 0
            }
            for draw in draws.split(";"):
                for cube in draw.split(","):
                    cube_count, cube_color = cube.strip().split(" ")
                    cubes_max[CubeColor(cube_color)] = max(cubes_max[CubeColor(cube_color)], int(cube_count))
            game_data[int(game_id)] = cubes_max
    return game_data


def solution_part_1(game_data: dict[int, dict[CubeColor, int]]):
    possible_games_sum = 0
    for index, game in game_data.items():
        if (
            game[CubeColor.RED] <= MAX_CUBES[CubeColor.RED] and
            game[CubeColor.GREEN] <= MAX_CUBES[CubeColor.GREEN] and
            game[CubeColor.BLUE] <= MAX_CUBES[CubeColor.BLUE]
        ):
            possible_games_sum += index
    print(possible_games_sum)


def solution_part_2(game_data: dict[int, dict[CubeColor, int]]):
    power_sum = 0
    for game in game_data.values():
        power_sum += game[CubeColor.RED] * game[CubeColor.GREEN] * game[CubeColor.BLUE]
    print(power_sum)


def main():
    solution_part_2(load_game("input.txt"))


if __name__ == "__main__":
    main()
